source/common/filter/tcp_proxy.h

```diff
@@ -199,5 +241,46 @@ class TcpProxy : public Network::ReadFilter,
   uint32_t connect_attempts_{};
 };
 
+// This class holds ownership of an upstream connection that needs to finish
+// flushing, when the downstream connection has been closed. The TcpProxy is
+// destroyed when the downstream connection is closed, so moving the upstream
+// connection here allows it to finish draining or timeout.
+class TcpProxyDrainer : public Event::DeferredDeletable {
+public:
+  TcpProxyDrainer(TcpProxyUpstreamDrainManager& parent,
+                  TcpProxyConfig::SharedConfigSharedPtr config,
+                  std::shared_ptr<TcpProxy::UpstreamCallbacks> callbacks,
+                  Network::ClientConnectionPtr connection, Event::TimerPtr idle_timer);
+
+  void onEvent(Network::ConnectionEvent event);
+  void onIdleTimeout();
+  void onBytesSent();
+  void cancelDrain();
+
+private:
+  TcpProxyUpstreamDrainManager& parent_;
+  std::shared_ptr<TcpProxy::UpstreamCallbacks> callbacks_;
+  Network::ClientConnectionPtr upstream_connection_;
+  Event::TimerPtr timer_;
+  TcpProxyConfig::SharedConfigSharedPtr config_;
+};
+
+typedef std::unique_ptr<TcpProxyDrainer> TcpProxyDrainerPtr;
+
+class TcpProxyUpstreamDrainManager : public ThreadLocal::ThreadLocalObject {
+public:
+  ~TcpProxyUpstreamDrainManager();
+  void add(TcpProxyConfig::SharedConfigSharedPtr config,
+           Network::ClientConnectionPtr upstream_connection,
+           std::shared_ptr<TcpProxy::UpstreamCallbacks> callbacks, Event::TimerPtr idle_timer);
+  void remove(TcpProxyDrainer& drainer, Event::Dispatcher& dispatcher);
+
+private:
+  // This must be a map instead of set because there is no way to move elements
+  // out of a set, and these elements get passed to deferredDelete() instead of
+  // being deleted in-place. The key and value will always be equal.
+  std::map<TcpProxyDrainer*, TcpProxyDrainerPtr> drainers_;
```

Comment:
Any reason not to use unordered_map here?

Reply:
Habit (not a good reason)